test(button-group): add a11y, visual tests

dancormier opened the pull request:
This PR adds visual and accessibility tests for the `s-btn-group` component.


> ~~**Note**~~
> ~~When constructing these tests, I realized we had used `--` to construct the `radio` modifier class (`.s-btn-group--radio`) instead of `__`. `--` is reserved for denoting _child_ elements of a component. This PR resolves that by changing the class to `.s-btn-group__radio`, which will result in a (easy to address) breaking change in the one place we use it: [stacks-editor](https://github.com/StackExchange/Stacks-Editor/blob/2124fd6ec1fb0b6569a4f9d2fea8686984120292/src/stacks-editor/editor.ts#L348).~~

~~@giamir since we'll need to update Core to support the [new button group design](https://github.com/StackExchange/Stacks/pull/1630), would this be a good opportunity to also include the minor class syntax breaking change noted above?~~

In the button group redesign implementation, the `radio` modifier` on `s-btn-group` no longer provides any unique styling. Consumers _should_ remove any references to this modifier, but including it will not affect anything materially.

dancormier committed:
test(button-group): add a11y, visual tests

dancormier committed:
fix(button-group): change s-btn-group--radio syntax to use `__`

dancormier committed:
found some lint

dancormier committed:
add static width to visual tests

dancormier committed:
update visual test shots

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Formatting!

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
fix(btn-group): add fix for buttons in forms (#1459)

* fix(btn-group): add fix for buttons in forms

* add tests (needs baseline images rendered

* update baseline images

dancormier committed:
formatting

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Remove test images to see if this makes it pass 🤷‍♂️

dancormier committed:
Reinstate visual test and baseline images

PLEASE DON'T FAIL!

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Update tests, baseline images

giamir commented on lib/components/button-group/button-group.a11y.test.ts:
```diff
@@ -0,0 +1,66 @@
+        options: {
+            includeNullModifier: false,
+        },
+        skippedTestids: ["s-btn-group-dark-radio"],
```
What's the problem with this variant?

dancormier replied:
I don't remember 😂 but it's no longer an issue with the updated design 🎉

giamir reviewed (changes requested):
Somehow the button group radio in the docs is now broken for me (cannot even focus on it). I haven't investigated further though.

For the tests I can see that the setup for the a11y and the visual tests is pretty similar. I would recommend to create something like `button-group.test.setup.ts` so that there is no need to duplicate many line of codes. You could also tests the `form` variations in the a11y tests if you want which probably makes sense (???).
If you go with the `.setup.ts` file approach we may want to add the `*.setup.ts` in[ this tsconfig](https://github.com/StackExchange/Stacks/blob/develop/lib/tsconfig.build.json#L3) to avoid creating unnecessary type for the production build.

I understand that we are introducing a small breaking change here both in Stacks Classic and Stacks Editor (which will depend on consumers upgrading the stacks classic css they use in their page).
Do you feel comfortable merging this change as part of v 2.1.0 or shall we wait for a future v2.2.0? I am asking because I would like to don't rush this through if possible due to enterprise release fixes we need to ship.

Apart from those minor things, the rest looks good. 
Thanks @dancormier for adding the tests 🎉

dancormier committed:
Merge branch 'develop' into dcormier/s-btn-group-tests

dancormier committed:
Update visual and a11y tests

dancormier committed:
Formatting

giamir commented:
> @giamir since we'll need to update Core to support the https://github.com/StackExchange/Stacks/pull/1630, would this be a good opportunity to also include the minor class syntax breaking change noted above?

~Yep, I think that makes sense. Let's make sure to call it out in the release notes.~

Class should be removed all together.

giamir commented on lib/components/button-group/button-group.a11y.test.ts:
```diff
@@ -0,0 +1,64 @@
+        },
+    });
+
+    // s-btn-group--radio
```
Should probably change to `s-btn-group__radio`

giamir commented on lib/components/button-group/button-group.a11y.test.ts:
```diff
@@ -0,0 +1,64 @@
+    isRadio?: boolean;
+    isSelected?: boolean;
+}): string => {
+    const baseClasses = "s-btn s-btn__muted s-btn__outlined";
```
I thought we don't need `s-btn__outlined` anymore

giamir commented on lib/components/button-group/button-group.visual.test.ts:
```diff
@@ -0,0 +1,137 @@
+    { name: "Unanswered" },
+];
+
+const form = (children: number[]) => `
```
We can remove those variants since they don't add much value, screenshots are identical

giamir commented on lib/components/button-group/button-group.visual.test.ts:
```diff
@@ -0,0 +1,137 @@
+        });
+    });
+
+    // s-btn-group--radio
```
Should probably change to `s-btn-group__radio`. Class will probably be removed all together.

giamir commented on docs/_data/button-groups.json:
```diff
@@ -6,7 +6,7 @@
     },
     {
-      "class": ".s-btn-group--radio",
+      "class": ".s-btn-group__radio",
```
Unused class. Removed from docs as agreed in Slack. :)

giamir commented on screenshots/Chromium/baseline/s-btn-group-dark.png:
This variation has a problem. Need investigation

giamir reviewed (changes requested):
Few changes to make before we are able to merge this one. 
Will circle back here when you finish the cleanup. :) 
Thanks

dancormier committed:
Simplify tests (round 1)

I removed any reference to the `s-btn-group__radio` variant and all but one `form` test

dancormier committed:
Move duplicated test code to setup file

As suggested in https://github.com/StackExchange/Stacks/pull/1346#pullrequestreview-1867360975

dancormier committed:
Simplify test elements

dancormier committed:
Remove reference to s-btn-group__radio from docs

dancormier committed:
tweak radio docs

dancormier committed:
add badges to tests, tweak buttons

dancormier commented:
@giamir I've updated this PR with your suggested changes. Since your last review I:

- Removed all references to `s-btn-group--radio` and `s-btn-group__radio`
- Abstracted most button group test code to `button-group.test.setup.ts`
- Removed all but one `form` test (tests that wrap a `button` with `form` element)
- Added tests for buttons with badges
- Simplified the children tested

> [!NOTE]
> The default, form, and radio test images render identically. This is as intended to test those unique children within the button group.

dancormier committed:
Remove `radio-with`badge` test variant

This seems unnecessary since the visual element is `.s-btn` regardless of whether it's a radio group or a standard button.

dancormier committed:
Add a11y test skipped test ids

dancormier committed:
Update skipped test comment with reference to btn test PR

giamir committed:
test(btn-group): remove unnecessary eslint disable rule

giamir reviewed (approved):
Outstanding work. Reads much better than before and great catch on those a11y failures. 🎉

I took the liberty to remove an unnecessary `eslint-disable` rule from the test setup. 
Good to be merged :)
